### PR TITLE
Add Flotiq and Astro integration guide

### PR DESCRIPTION
**Pull request by trzcina**
This documentation provides a step-by-step guide for integrating the [Flotiq](https://flotiq.com/) headless CMS with Astro. It helps developers enhance their Astro projects by incorporating dynamic content from Flotiq.

The guide covers the following topics:

- Prerequisites for setting up the integration.
- Creating and configuring content types in Flotiq.
- Connecting an Astro project to Flotiq's API.
- Fetching and displaying content from Flotiq using code examples.
- Handling both Static Site Generation (SSG) and Server-Side Rendering (SSR) modes.

### CODE REVIEW

**Comment by sarah11918** [comment]
Hi @trzcina ! Thank you for submitting this guide! We hope to be able to start reviewing it this week! :rocket:

**Review by TheOtterlord — COMMENTED** [review]
Thanks for the PR @trzcina! I've run through the recipe once, and made some initial notes about the order of steps, and using other package managers. Once we figure those out, I'll go through and review the copy.

The only issue I ran into was regenerating the SDK. I used `pnpm`, but running `pnpm add flotiq-api-ts`, even with `--force`, didn't regenerate the SDK after changing my content type. Deleting `node_modules` and installing with `npm` worked fine. Is it possible to regenerate the SDK with `pnpm` at all?

**Comment by trzcina** [comment]
Thanks @TheOtterlord! I’ll prepare an improved version incl. a component featuring tabs.

The most important thing for me is to check why the SDK didn’t update after `pnpm add` with `--force`. I’ll let you know.

File: src/content/docs/en/guides/cms/flotiq.mdx
Comment (trzcina) [reply src/content/docs/en/guides/cms/flotiq.mdx]: I have prepared an updated version; for pnpm, in this case, it's better to use `rebuild`. I’ve included this for Yarn, pnpm, and npm. Please check if this works for you.

**Comment by sarah11918** [comment]
Hey @trzcina ! Thanks for working with @TheOtterlord and for your patience! It looks like all the procedure was figured out, and I'll be looking at this in the next couple of days for a style/language pass. We should have this merged by the end of the week, if all goes well!  🙌

File: src/content/docs/en/guides/cms/flotiq.mdx
Comment (sarah11918) [inline src/content/docs/en/guides/cms/flotiq.mdx:282]:
``````suggestion
---
title: Flotiq & Astro
description: Add content to your Astro project using Flotiq as a CMS
type: cms
service: Flotiq
i18nReady: true
---

import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';

[Flotiq](https://flotiq.com) is a headless CMS designed for various frontends, such as static sites, mobile, and web applications. Developers and content creators manage and deliver content through REST and GraphQL-based APIs.

## Integrating with Astro

This guide will use the Flotiq headless CMS API with an Astro project to display content on your website.

### Prerequisites

To get started, you will need:

1. **An Astro project** - You can create a new project using the `npm create astro@latest` command.
2. **A Flotiq account** - If you don't have an account, [sign up for free](https://editor.flotiq.com/register).
3. **Flotiq read-only API key** - Find out [how to obtain your key](https://flotiq.com/docs/API/).

### Setting up the Environment variables

Add the read-only API key from your Flotiq account to the `.env` file in the root of your Astro project:

```ini title=".env"
FLOTIQ_API_KEY=__YOUR_FLOTIQ_API_KEY__
```

### Defining a Content Type in Flotiq

First, you need to define example [Content Type Definition](https://flotiq.com/docs/panel/content-types/) in Flotiq to store data.

Log in to your Flotiq account and create a custom Content Type Definition with the following example `Blog Post` configuration:
- **Label**: Blog Post
- **API Name**: blogpost
- **Fields**:
  - **Title**: text, required
  - **Slug**: text, required
  - **Content**: rich text, required

Then, create one or more example [Content Objects](https://flotiq.com/docs/panel/content-objects/) using this `Blog Post` type.

### Installing the Flotiq TypeScript SDK

To connect your project with Flotiq, install the [Flotiq SDK](https://github.com/flotiq/flotiq-api-ts) using the package manager of your choice:

<PackageManagerTabs>
  <Fragment slot="npm">
    ```sh
    npm install flotiq-api-ts
    ```
  </Fragment>
  <Fragment slot="pnpm">
    ```sh
    pnpm add flotiq-api-ts
    ```
  </Fragment>
  <Fragment slot="yarn">
    ```sh
    yarn add flotiq-api-ts
    ```
  </Fragment>
</PackageManagerTabs>

Next, configure the SDK using your credentials. Create a new file named `flotiq.ts` inside the `src/lib` directory of your project:

```ts title="src/lib/flotiq.ts"
import { FlotiqApi } from "flotiq-api-ts";

export const flotiq = new FlotiqApi(import.meta.env.FLOTIQ_API_KEY);
```

This configuration can now be used throughout your project.

### Fetching and Displaying Data from Flotiq

First, fetch the `Blog Post` data on an Astro page using your content's custom API `BlogpostAPI`:

```astro title="src/pages/index.astro"
---
import { flotiq } from "../lib/flotiq";

const posts = await flotiq.BlogpostAPI.list();
---
```

Then, display the content in your Astro template. You will have access to the `title`, `slug`, and `content` of your posts as well as other `internal` post data:

```astro title="src/pages/index.astro" ins={6-21}
---
import { flotiq } from "../lib/flotiq";

const posts = await flotiq.BlogpostAPI.list();
---
<html lang="en">
	<head>
		<title>Astro</title>
	</head>
	<body>
		{posts.data?.map((post) => (
			<section>
				<a href={`/posts/${post.slug}`}>
				  <h2>{post.title}</h2>
        </a>
				<div>{post.internal?.createdAt}</div>
				<div set:html={post.content}/>
			</section>
		))}
	</body>
</html>
```
Start the dev server and visit your page preview at `http://localhost:4321` to see the list of your blog posts. Each post will link to a page that does not yet exist. These will be created in the next step.

### Generating Individual Pages

Astro supports both prerendering all your pages ahead of time, or creating routes on demand when they are requested. Follow the instructions for either [static site generation](#static-site-generation) or [on-demand rendering](#on-demand-rendering) to build the page routes for your blog posts.

#### Static Site Generation

In static site generation (SSG) mode, use the `getStaticPaths()` method to fetch all possible blog post paths from Flotiq.

Create a new file `[slug].astro` in the `/src/pages/posts/` directory. Fetch all blog posts and return them within the `getStaticPaths()` method:

```astro title="src/pages/posts/[slug].astro" 
---
import type { Blogpost } from "flotiq-api-ts";
import { flotiq } from "../../lib/flotiq";

export async function getStaticPaths() {
  const posts = await flotiq.BlogpostAPI.list();
  return posts.data?.map((post) => ({
    params: { slug: post.slug },
    props: post
  })) || []
}
---
```

Next, add the templating to display an individual post:

```astro title="src/pages/posts/[slug].astro" ins={12-20}
---
import type { Blogpost } from "flotiq-api-ts";
import { flotiq } from "../../lib/flotiq";

export async function getStaticPaths() {
  const posts = await flotiq.BlogpostAPI.list();
  return posts.data?.map((post) => ({
    params: { slug: post.slug },
    props: post
  })) || []
}
const post: Blogpost = Astro.props;
---
<html lang="en">
  <title>{post.title}</title>
  <body>
    <h1>{post.title}</h1>
    <div set:html={post.content}/>
  </body>
</html>
```

Now when you visit `http://localhost:4321` and click on a blog post, you will be able to navigate to the individual post's page.

#### On-demand Rendering

If you are using [SSR](/en/guides/server-side-rendering/) mode, you will need to fetch a single post based on its `slug`.

Create a new file `[slug].astro` in the `/src/pages/posts/` directory. Fetch the post by its `slug` field, including logic to display a 404 page when the route is not found:

```astro title="src/pages/posts/[slug].astro"
---
import type { Blogpost } from "flotiq-api-ts";
import { flotiq } from "../../lib/flotiq";

const { slug } = Astro.params;
let post: Blogpost;

const blogpostList = await flotiq.BlogpostAPI.list({
  filters: JSON.stringify({
    slug: {
      type: 'equals',
      filter: slug,
    }
  }),
  limit: 1
});

if (blogpostList.data?.[0]) {
  post = blogpostList.data[0]
} else {
  return Astro.redirect('/404');
}
---
```

Next, add the templating to display an individual post:

```astro title="src/pages/posts/[slug].astro" ins={24-30}
---
import type { Blogpost } from "flotiq-api-ts";
import { flotiq } from "../../lib/flotiq";

const { slug } = Astro.params;
let post: Blogpost;

const blogpostList = await flotiq.BlogpostAPI.list({
  filters: JSON.stringify({
    slug: {
      type: 'equals',
      filter: slug,
    }
  }),
  limit: 1
});

if (blogpostList.data?.[0]) {
  post = blogpostList.data[0]
} else {
  return Astro.redirect('/404');
}
---
<html lang="en">
  <title>{post.title}</title>
  <body>
    <h1>{post.title}</h1>
    <div set:html={post.content}/>
  </body>
</html>
```

Now when you visit `http://localhost:4321` and click on a blog post, you will be able to navigate to the individual post's page.

### Refreshing the SDK After Content Type Changes

When using the Flotiq TypeScript SDK (`flotiq-api-ts`), all your data types are accurately mapped into the Astro project.

If you make changes to the structure of your content types (such as adding a new field or modifying an existing one), you’ll need to refresh the SDK to ensure that your project reflects the latest model updates.

To do this, run the rebuild command for your package manager:

<PackageManagerTabs>
  <Fragment slot="npm">
    ```sh
    npm rebuild flotiq-api-ts
    ```
  </Fragment>
  <Fragment slot="pnpm">
    ```sh
    pnpm rebuild flotiq-api-ts
    ```
  </Fragment>
  <Fragment slot="yarn">
    ```sh
    yarn add flotiq-api-ts
    ```
  </Fragment>
</PackageManagerTabs>

This will update the SDK, aligning object types, fields, and API methods with your current data model.


## Publishing Your Site

To deploy your website, visit Astro's [deployment guides](/en/guides/deploy/) and follow the instructions for your preferred hosting provider.

### Redeploy on Flotiq Changes

To update your published site, configure Flotiq to send a webhook your hosting provider to trigger a rebuild whenever your content changes.

In Flotiq, you can define which Content Type and events it should trigger on, and configure it accordingly. See the [Flotiq Webhooks documentation](https://flotiq.com/docs/panel/webhooks/async-co-webhook/) for more details.

## Official Resources

- [Flotiq documentation](https://flotiq.com/docs)
``````

**Review by sarah11918 — COMMENTED** [review]
Sorry, you didn't give permission for maintainers to edit, so I had to make one big suggestion on the whole file!  😅 

I will commit this, then ask you to review the light text edits I've made!

**Comment by trzcina** [comment]
Thanks, @sarah11918! I reviewed and committed your suggested changes. Thanks for the suggestions — looks good now.

**Comment by sarah11918** [comment]
Hi @trzcina ! Thank you for making that commit, but you also have to give maintainers permission to edit so we can do basic things like update the branch.

Please edit the permissions on this PR for us so we can continue with it! :raised_hands: Then it won't be too long until we can merge it! :partying_face:

File: src/content/docs/en/guides/cms/flotiq.mdx
Comment (casungo) [inline src/content/docs/en/guides/cms/flotiq.mdx:260]: shouldn't this be rebuild? or am i missing something

```suggestion
    yarn rebuild flotiq-api-ts
```

**Review by casungo — COMMENTED** [review]
LGTM! just a quick note

**Comment by trzcina** [comment]
> Hi @trzcina ! Thank you for making that commit, but you also have to give maintainers permission to edit so we can do basic things like update the branch.
> 
> Please edit the permissions on this PR for us so we can continue with it! 🙌 Then it won't be too long until we can merge it! 🥳

Hi @sarah11918! I tried enabling the option to allow changes to a pull request branch created from a fork, but it looks like that option isn’t available in my PR (see this [discussion](https://github.com/orgs/community/discussions/5634)).

Instead, I’ve granted access to the repo for everyone who’s been working on this PR so far. Hopefully, that works for you! :)

File: src/content/docs/en/guides/cms/flotiq.mdx
Comment (trzcina) [reply src/content/docs/en/guides/cms/flotiq.mdx:260]: Thanks for the suggestion! The proposed `yarn add` works and rebuilds the package each time (which is what we expect), but this applies specifically to yarn v1. In yarn v4, the appropriate command for this is `yarn rebuild`.

I suggest updating the line to `yarn rebuild flotiq-api-ts` (as you suggested) and optionally adding a note:
"If you are using yarn v1 (Classic), run `yarn add flotiq-api-ts` to rebuild the SDK."

**Comment by sarah11918** [comment]
Ah, interesting! I wasn't aware of the current limitation. It certainly makes it more annoying as the maintainer of a PR that we can't handle routine things on our end. 😂 

I saw the invitation, and thank you! I don't see why this shouldn't work, but I'll let you know if I have any trouble. 💜

File: src/content/docs/en/guides/cms/flotiq.mdx
Comment (casungo) [reply src/content/docs/en/guides/cms/flotiq.mdx:260]: Yes, it can be done! Perhaps let's get another opinion on this, but what you proposed works for me. It will help avoid confusion for the end user who sees one method for all the other package managers and a different one for Yarn.

**Review by sarah11918 — APPROVED** [review]
LGTM! Thank you everyone, and welcome to Team Docs, @trzcina ! :partying_face:

**Pull request merged**
